tools/nnyeah/nnyeah/Reworker.cs

```diff
@@ -0,0 +1,481 @@
+				ReworkMethod (method);
+			}
+		}
+
+		void ReworkMethod (MethodDefinition method)
+		{
+			var nativeTypes = new List<bool> ();
+			var changed = false;
+			foreach (var parameter in method.Parameters) {
+				nativeTypes.Clear ();
+				changed = false;
+				var newType = ReworkTypeReference (parameter.ParameterType, nativeTypes, ref changed);
+				if (changed) {
+					parameter.ParameterType = newType;
+				}
+				if (nativeTypes.Contains (true)) {
+					parameter.CustomAttributes.Add (NativeIntAttribute (nativeTypes));
+				}
+			}
+
+			nativeTypes.Clear ();
+			changed = false;
+			var newReturnType = ReworkTypeReference (method.ReturnType, nativeTypes, ref changed);
+			if (changed) {
+				method.ReturnType = newReturnType;
+			}
+			if (nativeTypes.Contains (true)) {
+				method.MethodReturnType.CustomAttributes.Add (NativeIntAttribute (nativeTypes));
+			}
+
+			if (method.Body != null) {
+				foreach (var variable in method.Body.Variables) {
+					nativeTypes.Clear ();
+					changed = false;
+					var newType = ReworkTypeReference (variable.VariableType, nativeTypes, ref changed);
+					if (changed) {
+						variable.VariableType = newType;
+					}
+				}
+				ReworkCodeBlock (method.Body);
+			}
+		}
+
+		void ReworkProperty (PropertyDefinition prop)
+		{
+			var nativeTypes = new List<bool> ();
+			var changed = false;
+			var newPropType = ReworkTypeReference (prop.PropertyType, nativeTypes, ref changed);
+			if (changed) {
+				prop.PropertyType = newPropType;
+			}
+			if (nativeTypes.Contains (true)) {
+				prop.CustomAttributes.Add (NativeIntAttribute (nativeTypes));
+			}
+			ReworkMethods (PropMethods (prop));
+		}
+
+		void ReworkField (FieldDefinition field)
+		{
+			var nativeTypes = new List<bool> ();
+			var changed = false;
+			var newType = ReworkTypeReference (field.FieldType, nativeTypes, ref changed);
+			if (changed) {
+				field.FieldType = newType;
+			}
+			if (nativeTypes.Contains (true)) {
+				field.CustomAttributes.Add (NativeIntAttribute (nativeTypes));
+			}
+		}
+
+		void ReworkEvent (EventDefinition @event)
+		{
+			var nativeTypes = new List<bool> ();
+			var changed = false;
+			var newType = ReworkTypeReference (@event.EventType, nativeTypes, ref changed);
+			if (changed) {
+				@event.EventType = newType;
+			}
+			if (nativeTypes.Contains (true)) {
+				@event.CustomAttributes.Add (NativeIntAttribute (nativeTypes));
+			}
+			ReworkMethods (EventMethods (@event));
+		}
+
+		CustomAttribute NativeIntAttribute (List<bool> nativeTypes)
+		{
+			var nativeIntAttr = new CustomAttribute (new MethodReference (".ctor", module.TypeSystem.Void, nativeIntegerAttributeTypeRef));
+			if (nativeTypes.Count > 1) {
+				var boolArrayParameter = new ParameterDefinition (module.TypeSystem.Boolean.MakeArrayType ());
+				nativeIntAttr.Constructor.Parameters.Add (boolArrayParameter);
+				var boolArray = nativeTypes.Select (b =>
+					new CustomAttributeArgument (module.TypeSystem.Boolean, b)).ToArray ();
```

Comment:
Does it need to be an Array? Can it be an IEnumerable so that we are lazy?

Reply:
It cannot be IEnumerable. The boolArray really needs to be an array for attribute constructor. The attribute is something that in dotnet 6 gets injected into the assembly by the compiler. Pre dotnet 6 won't have it, so we have to inject it into the code and we have to abide by its API which is a bool array.

Reply:
Can you show me an example when it is not readable, maybe we can do something about it.